zmq/decorators/base.py

```diff
@@ -0,0 +1,62 @@
+from functools import wraps
+
+from zmq.utils.strtypes import basestring
+
+
+class ZDecoratorBase(object):
+
+    def __init__(self, target):
+        self.target = target
+
+    def __call__(self, *dec_args, **dec_kwargs):
+        self.kwname = None
+        self.dec_args = dec_args
+        self.dec_kwargs = dec_kwargs
+
+        self.pop_kwname()
+
+        def decorator(func):
+            @wraps(func)
+            def wrapper(*wrap_args, **wrap_kwargs):
+                self.wrap_args = wrap_args
+                self.wrap_kwargs = wrap_kwargs
+
+                self.hook('preinit')
+
+                with self.target(*self.dec_args, **self.dec_kwargs) as obj:
+                    self.hook('postinit')
+
+                    if self.kwname and self.kwname not in self.wrap_kwargs:
+                        self.wrap_kwargs[self.kwname] = obj
+                    elif self.kwname and self.kwname in self.wrap_kwargs:
+                        raise TypeError(
+                            "{0}() got multiple values for"
+                            " argument '{1}'".format(
+                                func.__name__, self.kwname))
+                    else:
+                        self.wrap_args = (obj,) + self.wrap_args
+
+                    self.hook('preexec')
+                    func(*self.wrap_args, **self.wrap_kwargs)
+                    self.hook('postexec')
+
+                self.hook('cleanup')
```

Comment:
I place hooks here, and let the child class can use it.
I add more doc strings for `ZDecoratorBase`, please check out 10c33b9.